python/ray/import_thread.py

```diff
@@ -0,0 +1,209 @@
+        self.worker = worker
+        self.mode = mode
+        self.redis_client = worker.redis_client
+
+    def start(self):
+        """Start the import thread."""
+        t = threading.Thread(target=self._run)
+        # Making the thread a daemon causes it to exit
+        # when the main thread exits.
+        t.daemon = True
+        t.start()
+
+    def _run(self):
+        from ray.worker import profile, WORKER_MODE
+        import_pubsub_client = self.redis_client.pubsub()
+        # Exports that are published after the call to
+        # import_pubsub_client.subscribe and before the call to
+        # import_pubsub_client.listen will still be processed in the loop.
+        import_pubsub_client.subscribe("__keyspace@0__:Exports")
+        # Keep track of the number of imports that we've imported.
+        num_imported = 0
+
+        # Get the exports that occurred before the call to subscribe.
+        with self.worker.lock:
+            export_keys = self.redis_client.lrange("Exports", 0, -1)
+            for key in export_keys:
+                num_imported += 1
+
+                # Handle the driver case first.
+                if self.mode != WORKER_MODE:
+                    if key.startswith(b"FunctionsToRun"):
+                        self.fetch_and_execute_function_to_run(key)
+                    # Continue because FunctionsToRun are the only things that
+                    # the driver should import.
+                    continue
+
+                if key.startswith(b"RemoteFunction"):
+                    self.fetch_and_register_remote_function(key)
+                elif key.startswith(b"FunctionsToRun"):
+                    self.fetch_and_execute_function_to_run(key)
+                elif key.startswith(b"ActorClass"):
+                    # Keep track of the fact that this actor class has been
+                    # exported so that we know it is safe to turn this worker
+                    # into an actor of that class.
+                    self.worker.imported_actor_classes.add(key)
+                else:
+                    raise Exception("This code should be unreachable.")
+
+        try:
+            for msg in import_pubsub_client.listen():
+                with self.worker.lock:
+                    if msg["type"] == "subscribe":
+                        continue
+                    assert msg["data"] == b"rpush"
+                    num_imports = self.redis_client.llen("Exports")
+                    assert num_imports >= num_imported
+                    for i in range(num_imported, num_imports):
+                        num_imported += 1
+                        key = self.redis_client.lindex("Exports", i)
+
+                        # Handle the driver case first.
+                        if self.mode != WORKER_MODE:
+                            if key.startswith(b"FunctionsToRun"):
+                                with profile(
+                                        "ray:import_function_to_run",
```

Comment:
I think the test failure in the "xray" Travis job is caused by the name changes in these `profile` statements. This PR changes them accidentally (due to the rebase). Do you see what I mean?

Reply:
I see. thank you for pointing out.